### PR TITLE
refactor(lsp): overhaul tsc request handling

### PR DESCRIPTION
**Pull request by lucacasonato**
This PR changes how the LSP interacts with the LSP. It implements prioritaztion
of interactive TSC requests over diagnostics, and prevents diagnostic requests
being queued in the incoming mpsc.

### CODE REVIEW

**Comment by lucacasonato** [comment]
This request has no benefits right now because the LSP handles all requests fully synchronously (https://github.com/denoland/deno/pull/9271). I was not aware that we lock up the entirety of the LSP for the entire duration of all incoming requests. This needs to be resolved first.

File: cli/lsp/diagnostics.rs
Comment (kitsonk) [inline cli/lsp/diagnostics.rs:232]: Maybe turn this into a proper docblock comment?  Otherwise it feels just like a stray comment.

File: cli/lsp/diagnostics.rs
Comment (kitsonk) [inline cli/lsp/diagnostics.rs:237]: Have you considered how this is handled when diagnostics are "cleared"?

Comment (kitsonk) [reply cli/lsp/diagnostics.rs:237]: I guess it only happens when they have been cancelled... right?

File: cli/lsp/tsc.rs
Comment (kitsonk) [inline cli/lsp/tsc.rs:75]:
```suggestion
    let moved_diagnostics_request = diagnostics_request.clone();
```

Or anything than a trailing `_` as I personally think it is just ambiguous what it represented. `diagnostics_request_clone` or whatever...

File: cli/lsp/tsc.rs
Comment (kitsonk) [inline cli/lsp/tsc.rs:98]:
```suggestion
              warn!("Internal Error: Unable to send tsc result to language server.");
```

**Review by kitsonk — APPROVED** [review]
A couple thoughts... if you rebase off main as well, we should get to see the benchmarks in the output of the CI.

File: cli/lsp/diagnostics.rs
Comment (lucacasonato) [reply cli/lsp/diagnostics.rs:237]: If they are cleared, a `Some` of an empty `DiagnosticVec` would be returned I think.

**Comment by lucacasonato** [comment]
I want to hold this until #9557 is landed. This PR alone will have no impact on performance.

**Pull request closed without merging**
